satellite/gc/observer.go

```diff
@@ -0,0 +1,82 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package gc
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/zap"
+
+	"storj.io/storj/pkg/bloomfilter"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+)
+
+// Observer implements the observer interface for gc
+type Observer struct {
+	log          *zap.Logger
+	pieceCounts  map[storj.NodeID]int
+	retainInfos  map[storj.NodeID]*RetainInfo
+	config       Config
+	creationDate time.Time
+}
+
+// NewObserver instantiates a gc Observer
+func NewObserver(log *zap.Logger, pieceCounts map[storj.NodeID]int, config Config) *Observer {
+	return &Observer{
+		log:          log,
+		pieceCounts:  pieceCounts,
+		retainInfos:  make(map[storj.NodeID]*RetainInfo),
+		config:       config,
+		creationDate: time.Now().UTC(),
+	}
+}
+
+// RemoteSegment takes a remote segment found in metainfo and adds pieces to bloom filters
+func (observer *Observer) RemoteSegment(ctx context.Context, path storj.Path, pointer *pb.Pointer) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	remote := pointer.GetRemote()
+	pieces := remote.GetRemotePieces()
+
+	for _, piece := range pieces {
+		pieceID := remote.RootPieceId.Derive(piece.NodeId, piece.PieceNum)
+		observer.add(ctx, piece.NodeId, pieceID)
+	}
+	return nil
+}
+
+// RemoteObject returns nil because gc does not interact with remote objects
+func (observer *Observer) RemoteObject(ctx context.Context, path storj.Path, pointer *pb.Pointer) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	return nil
+}
+
+// InlineSegment returns nil because we're only doing gc for storage nodes for now
+func (observer *Observer) InlineSegment(ctx context.Context, path storj.Path, pointer *pb.Pointer) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	return nil
+}
+
+// adds a pieceID to the relevant node's RetainInfo
+func (observer *Observer) add(ctx context.Context, nodeID storj.NodeID, pieceID storj.PieceID) {
+	var filter *bloomfilter.Filter
+
+	if _, ok := observer.retainInfos[nodeID]; !ok {
+		// If we know how many pieces a node should be storing, use that number. Otherwise use default.
+		numPieces := int(observer.config.InitialPieces)
+		if observer.pieceCounts[nodeID] > 0 {
+			numPieces = observer.pieceCounts[nodeID]
+		}
+		filter = bloomfilter.NewOptimal(numPieces, observer.config.FalsePositiveRate)
```

Comment:
this must have a size limit, otherwise the grpc call can fail

Reply:
can we use streaming?

Reply:
I don't think the bloom filters should be large enough that we will need streaming. But I am not sure how to calculate bloom filter size

Reply:
either way, what should the size limit be?

Reply:
max 2MB, default grpc limit is 4MB